### PR TITLE
Add user agreements attribute to userfas

### PR DESCRIPTION
**Pull request by ryanlerch**
Related: fedora-infra/noggin#51 and
https://github.com/fedora-infra/noggin/pull/270

### CODE REVIEW

**Comment by Conan-Kudo** [comment]
Question, why is this an attribute and not using groups? Most of our extensions and APIs assume that the user is a member of the group that signed the FPCA. Also, how would this extend to _multiple_ user agreements (e.g. code of conduct, then FPCA, etc.)?

**Comment by tiran** [comment]
> Question, why is this an attribute and not using groups? Most of our extensions and APIs assume that the user is a member of the group that signed the FPCA. Also, how would this extend to _multiple_ user agreements (e.g. code of conduct, then FPCA, etc.)?

It's a multi-valued string attribute. You can have multiple agreements.

**Comment by Conan-Kudo** [comment]
@tiran So why attribute and not groups?

File: updates/99-fas.update [hunk 1]
```diff
@@ -14,11 +14,11 @@ only:ipaUserAuthType: otp
 
 # ACI to expose FAS user attributes
 dn: cn=users,cn=accounts,$SUFFIX
-remove:aci: (targetattr = "nsAccountLock || fasTimezone || fasLocale || fasIRCNick || fasGPGKeyId || fasCreationTime || fasStatusNote")(targetfilter = "(objectclass=fasuser)")(version 3.0;acl "Read FAS user attributes";allow (compare,read,search) userdn = "ldap:///all";)
```
Comment (tiran) [inline updates/99-fas.update:17]: You have to keep the old removes to support migration from old plugin versions.

File: updates/99-fas.update [hunk 2]
```diff
@@ -14,11 +14,11 @@ only:ipaUserAuthType: otp
...
 
-remove:aci: (targetattr = "fasTimezone || fasLocale || fasIRCNick || fasGPGKeyId")(version 3.0;acl "Users can modify their own FAS attributes";allow (write) userdn = "ldap:///self";)
-add:aci: (targetattr = "fasTimezone || fasLocale || fasIRCNick || fasGPGKeyId || fasRHBZEmail || fasGitHubUsername || fasGitLabUsername")(version 3.0;acl "Users can modify their own FAS attributes";allow (write) userdn = "ldap:///self";)
+remove:aci: (targetattr = "fasTimezone || fasLocale || fasIRCNick || fasGPGKeyId || fasRHBZEmail || fasGitHubUsername || fasGitLabUsername")(version 3.0;acl "Users can modify their own FAS attributes";allow (write) userdn = "ldap:///self";)
+add:aci: (targetattr = "fasTimezone || fasLocale || fasIRCNick || fasGPGKeyId || fasRHBZEmail || fasGitHubUsername || fasGitLabUsername || fasUserAgreements")(version 3.0;acl "Users can modify their own FAS attributes";allow (write) userdn = "ldap:///self";)
```
Comment (tiran) [inline updates/99-fas.update:21]: Is it a good idea to allow users to modify the attribute without checks? This allows a user to drop an agreement or add custom agreements, too.

**Review by tiran — CHANGES_REQUESTED** [review]

**Comment by tiran** [comment]
> @tiran So why attribute and not groups?

Why do you insist on groups? I was under the impression that agreement groups in old FAS were a workaround to simplify implementation, not an inherit design feature.

A separate attribute allows for different access rights.

**Comment by Conan-Kudo** [comment]
@tiran Mainly because I wanted to know and I didn't know the background for the groups from before.

**Comment by tiran** [comment]
> @tiran Mainly because I wanted to know and I didn't know the background for the groups from before.

That's always a good reason to ask :+1:

IMO it's a good idea to treat agreements and groups differently. FAS does not allow users to join a group easily but allows users to leave a group on request. Agreements should work the other way around or allow for more fine-grained access control. I proposed a solution that is more complex but allows for more control in the [ticket](https://github.com/fedora-infra/noggin/issues/51).

**Comment by ryanlerch** [comment]
@tiran yeah, the more complex solution is a go in my book.

closing this PR

**Pull request closed without merging**